x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/RandomizedAssignmentRounding.java

```diff
@@ -0,0 +1,428 @@
+                }
+                int extraAllocations = Math.min(
+                    resourceTracker.remainingNodeCores.get(n) / m.threadsPerAllocation(),
+                    resourceTracker.remainingModelAllocations.get(m)
+                );
+                softAllocations.compute(Tuple.tuple(m, n), (k, v) -> v + extraAllocations);
+                resourceTracker.assign(m, n, extraAllocations);
+            }
+
+            zeroSoftAssignmentsOfSatisfiedModels();
+        }
+
+        private double remainingModelOrder(Model m) {
+            return (m.currentAllocationByNodeId().isEmpty() ? 1 : 2) * -m.memoryBytes();
+        }
+
+        private boolean hasSoftAssignments(Node n) {
+            return models.stream().anyMatch(m -> isSoftAssignment(m, n));
+        }
+
+        private boolean isSoftAssignment(Model m, Node n) {
+            Tuple<Model, Node> index = Tuple.tuple(m, n);
+            return (softAssignments.get(index) > 0 && softAssignments.get(index) < 1) || isInteger(softAllocations.get(index)) == false;
+        }
+
+        private void zeroSoftAssignmentsOfSatisfiedModels() {
+            for (Model m : models) {
+                if (resourceTracker.remainingModelAllocations.get(m) <= 0) {
+                    for (Node n : nodes) {
+                        Tuple<Model, Node> index = Tuple.tuple(m, n);
+                        if (isSoftAssignment(m, n)) {
+                            unassign(index);
+                        }
+                    }
+                }
+            }
+        }
+
+        private void unassign(Tuple<Model, Node> assignment) {
+            softAssignments.put(assignment, 0.0);
+            softAllocations.put(assignment, 0.0);
+        }
+
+        private List<Tuple<Model, Node>> createSoftAssignmentQueue() {
+            List<Tuple<Model, Node>> queue = new ArrayList<>();
+            models.forEach(m -> nodes.forEach(n -> {
+                if (isSoftAssignment(m, n)) {
+                    queue.add(Tuple.tuple(m, n));
+                }
+            }));
+            queue.sort(
+                Comparator.comparingDouble(this::assignmentDistanceFromZeroOrOneOrder)
+                    .thenComparingDouble(this::assignmentMostRemainingThreadsOrder)
+            );
+            return queue;
+        }
+
+        private double assignmentDistanceFromZeroOrOneOrder(Tuple<Model, Node> assignment) {
+            return Math.min(softAssignments.get(assignment), 1 - softAssignments.get(assignment));
+        }
+
+        private double assignmentMostRemainingThreadsOrder(Tuple<Model, Node> assignment) {
+            return -softAllocations.get(assignment) * assignment.v1().threadsPerAllocation();
+        }
+
+        private void doRandomizedRounding(List<Tuple<Model, Node>> softAssignmentQueue) {
+            for (Tuple<Model, Node> assignment : softAssignmentQueue) {
+                if (softAssignments.get(assignment) == 1 && isInteger(softAllocations.get(assignment))) {
```

Comment:
I think this warrants a comment: i.e. because other operations can snap nodes in the queue as a side effect. Also shouldn't this be
```suggestion
                if (isSoftAssignment(m, n) == false) {
                    ...
```
and just after unpacking the tuple.